src/instrumenter/state_var_instrumenter.ts

```diff
@@ -0,0 +1,1165 @@
+import { assert } from "../util";
+import { InstrumentationContext } from "./instrumentation_context";
+import { parse as parseTypeString } from "../spec-lang/typeString_parser";
+import { TranspilingContext } from "./transpiling_context";
+import {
+    SAddressType,
+    SArrayType,
+    SBoolType,
+    SBytes,
+    SFixedBytes,
+    SId,
+    SIfUpdated,
+    SIntLiteralType,
+    SIntType,
+    SMappingType,
+    SPointer,
+    SStateVarProp,
+    SString,
+    STupleType,
+    SType,
+    SUserDefinedType
+} from "../spec-lang/ast";
+import { SStringLiteralType } from "../spec-lang/ast/types/string_literal";
+import { astTypeNameToSType } from "../spec-lang/tc";
+import { makeTypeString } from "./type_string";
+import { cook } from "../rewriter";
+import { getOrAddConstructor } from "./instrument";
+import { lt } from "semver";
+
+/**
+ * Given a Solidity `Expression` `e` and the `expectedType` where its being used,
+ * compute the actual `SType` of `e`. This may be different from `expectedType` due to implicit
+ * casts.
+ *
+ * Note: We only need `expectedType` here due to limitation in typeString parsing in the case
+ * of int literals. This function should be removed after we add "get type of arbitrary expression" to
+ * solc-typed-ast.
+ *
+ * @param e
+ * @param expectedType
+ */
+function getExprSType(e: Expression, expectedType?: SType): SType {
+    /**
+     * Sanitize the parsed `actualType` by replacing any int_const types with the
+     * concrete integer type expected at that location, and any string literal types with
+     * `string memory`. Note this code makes the assumption that int literals and string literal
+     * types CANNOT show up inside array/mapping types (which I think is true?).
+     *
+     * @param actualType
+     * @param expectedType
+     */
+    const sanitizeType = (actualType: SType, expectedType: SType): SType => {
+        if (actualType instanceof SIntLiteralType) {
+            assert(
+                expectedType instanceof SIntType,
+                `Expected ${expectedType.pp()} got ${actualType.pp()}`
+            );
+            return expectedType;
+        }
+
+        if (actualType instanceof SStringLiteralType) {
+            return new SPointer(new SString(), DataLocation.Memory);
+        }
+
+        if (actualType instanceof STupleType) {
+            assert(
+                expectedType instanceof STupleType &&
+                    expectedType.elements.length === actualType.elements.length,
+                `Expected ${expectedType.pp()} got ${actualType.pp()}`
+            );
+
+            return new STupleType(
+                actualType.elements.map((el, i) => sanitizeType(el, expectedType.elements[i]))
+            );
+        }
+
+        return actualType;
+    };
+
+    const parsedType: SType = parseTypeString(e.typeString);
+
+    return expectedType !== undefined ? sanitizeType(parsedType, expectedType) : parsedType;
+}
+
+export function getKeysAndCompTypes(
```

Comment:
Hi, What does `Comp` mean here? path seems denote index and member accesses so I don't understand what `Comp` means here.

Reply:
Seems like `Comp` stands for `component` of `ConcreteDatastructurePath`. See [this loop construct](https://github.com/ConsenSys/scribble/blob/e7795b23fdcbd4c829f674b1ec3fee797b577e01/src/instrumenter/state_var_instrumenter.ts#L143).